node/core/candidate-validation/src/lib.rs

```diff
@@ -569,9 +569,13 @@ async fn validate_candidate_exhaustive(
 			Ok(ValidationResult::Invalid(InvalidCandidate::ExecutionError(
 				"ambiguous worker death".to_string(),
 			))),
-		Err(ValidationError::InvalidCandidate(WasmInvalidCandidate::PrepareError(e))) =>
-			Ok(ValidationResult::Invalid(InvalidCandidate::ExecutionError(e))),
-
+		Err(ValidationError::InvalidCandidate(WasmInvalidCandidate::PrepareError(e))) => {
```

Comment:
Actually now that I revisit this, this should be fixed at one layer down already: `WasmInvalidCandidate` is already the wrong error. The candidate is not at fault if preparation failed.

Reply:
If this is merged, actually more cleanup will be needed: 

[`PrecheckOutcome::Failed`](https://github.com/paritytech/polkadot/blob/dfa8c5ae321e0523f88fdc43afc6d743812ee49e/node/subsystem-types/src/messages.rs#L97) and the related [`PrepareError`](https://github.com/paritytech/polkadot/blob/be487ae3f8b3fda3cb0a2efe27b1500b550d1ae4/node/core/pvf/src/error.rs#L133).

Reason we no longer differentiate. During pre-checking we always assume invalid, just to be on the safe side. During preparation, we never want to raise a dispute if preparation fails (as it previously passed pre-checking), so we never want it to be of state `InvalidCandidate`.

Reply:
Differentiation between the two error types (deterministic and non deterministic) still makes sense. For deterministic errors happening in preparation, we should print a big fat error.

Reply:
> Differentiation between the two error types (deterministic and non deterministic) still makes sense. For deterministic errors happening in preparation, we should print a big fat error.

You can call `is_deterministic()` on `PrepareError`. (Semantically it made sense to me to have separate error types, but others didn't think it was necessary.)

Reply:
Yes. I arrived at the conclusion that the distinction is very useful. Non spurious errors, suggest some major issue like a hardware fault - the very least, they should have a higher log level.